cosmos/ethermint
# Implement eth_syncing

- Completes #21 eth_syncing

Node: Will add an issue for adding the other fields if they become available through Tendermint

Currently the endpoint works and checks correctly if the Tendermint node is "CatchingUp" which is the equivalent of syncing in the context of what it's used for in this rpc endpoint. WIP because the Tendermint status does not return starting or highest block which is defined in the spec as being the required return parameters when syncing.

Here is a dump of the data that can is retrieved through the Tendermint status query:
```
{"jsonrpc":"2.0","id":68,"result":{"node_info":{"protocol_version":{"p2p":7,"block":10,"app":0},"id":"62c234b460b678690c2be778619ff45c240cdbdc","listen_addr":"tcp://0.0.0.0:26656","network":"testchain","version":"0.32.2","channels":"4020212223303800","moniker":"moniker","other":{"tx_index":"on","rpc_address":"tcp://127.0.0.1:26657"}},"sync_info":{"latest_block_hash":"D8E298EFD8C64E292109F7D6BE4AD589C256A90E752734705DA3883A963F3A49","latest_app_hash":"608FA2651F664B85D64AC4902700B0CC9DA21E40C9E8105B58713E752DFF7490","latest_block_height":3,"latest_block_time":"2019-08-28T18:50:12.710709Z","catching_up":false},"validator_info":{"address":"F71E9ECA64BF3404C6001842525313CD43BF6EB7","pub_key":[151,96,251,213,194,181,146,155,2,25,168,0,173,99,25,39,22,156,3,36,29,16,11,253,221,183,222,215,144,45,207,211],"voting_power":100}}}
```

But currently the endpoint correctly checks if the node is syncing or not, just those two data points need to be pulled from somewhere to make it fully compliant. Getting number of state entries pulled and how many remain would be nice, but I don't think is required and would be a tedious workaround to get this data output.

To test:
1. Run the Ethermint node how it is normally run, I ran:
```
make install 
rm -rf ~/.emint*
emintd init moniker --chain-id testchain
emintcli config chain-id testchain
emintcli config output json
emintcli config indent true
emintcli config trust-node true
emintcli keys add austin
<password>
<password>
emintd add-genesis-account $(emintcli keys show austin -a) 1000photon,100000000stake
emintd gentx --name austin
<password>
emintd collect-gentxs
emintd validate-genesis
emintd start
```

2. Start rest server:
```
emintcli rest-server --laddr "tcp://localhost:8545"
```

3. Send syncing request:
```
curl -X POST --data '{"jsonrpc":"2.0","method":"eth_syncing","params":[],"id":68}' -H "Content-Type: application/json" http://localhost:8545/rpc
```

Edit: response should be:
```
{"jsonrpc":"2.0","id":68,"result":false}
```

running this since there will be no connected peers to connect to and catch up state through syncing.
